src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,212 @@
+  }
+
+  public getCardSize() {
+    return (this.config ? (this.config.title ? 1 : 0) : 0) + (this.items ? this.items.length : 0);
+  }
+
+  public setConfig(config: Config) {
+    this.config = config;
+    this.items = [];
+    this._fetchData();
+  }
+
+  public connectedCallback() {
+    if (!this.hass) {
+      return;
+    }
+
+    super.connectedCallback();
+    this.hass.connection
+      .subscribeEvents(this._fetchData, "shopping_list_updated")
+      .then(() => this._fetchData());
+  }
+
+  public disconnectedCallback() {
+    super.disconnectedCallback();
+  }
+
+  protected render() {
+    if (!this.config || !this.hass) {
+      return html``;
+    }
+
+    // TODO placeholder='${this.localize('ui.panel.shopping-list.add_item')}' is coming back with an empty value
+    return html`
+      ${this.renderStyle()}
+      <ha-card .header="${this.config.title}">
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='Add Me'
+            @keydown='${(e) => this._addKeyPress(e)}'
+            no-label-float
+          ></paper-input>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click='${(e) => this._addItem(e)}'
+          ></paper-icon-button>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:notification-clear-all"
+            @click='${() => this._clearItems()}'
+          ></paper-icon-button>
+        </div>
+
+      ${this.items!.map((item) => html`
+        <paper-icon-item>
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            @click='${(e) => this._itemCompleteTapped(e, item.id)}'
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
+              id='editBox'
+              no-label-float
+              .value='${item.name}'
+              @change='${(e) => this._saveEdit(e, item.id)}'
+            ></paper-input>
+          </paper-item-body>
+        </paper-icon-item>
+        `)}
+      </ha-card>
+    `;
+  }
+
+  private renderStyle() {
+    return html`
+      <style>
+        .addRow {
+          padding-left: 16px;
+          display: inline-block;
+        }
+        .addRow > paper-icon-button {
+          display: inline-block;
+        }
+        .addRow > paper-input {
+          display: inline-block;
+        }
+        paper-icon-item {
+          border-top: 1px solid var(--divider-color);
+        }
+        paper-icon-item:first-child {
+          border-top: 0;
+        }
+        paper-checkbox {
+          padding: 11px;
+        }
+        paper-input {
+          --paper-input-container-underline: {
+            display: none;
+          }
+          --paper-input-container-underline-focus: {
+            display: none;
+          }
+          position: relative;
+          top: 1px;
+        }
+      </style>
+    `;
+  }
+
+  private _fetchData() {
```

Comment:
You can use an async function instead of promises.

```js
private async _fetchData() {
  if (this.hass) {
    this._assignItems(await fetchItems(this.hass));
  }
}
```